# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/2d/DrawTargetD2D1.cpp
# Commit: ab254a4faafe
# Full Hash: ab254a4faafe7b1ada5701b293187996711c4138
# Author: Bas Schouten <bschouten@mozilla.com>
# Date: 2019-02-10 21:38:44
# Regressor Bug: 1521774
# File Overlap Count: 1
# Description:
#   Bug 1521774: Ensure a DC is available when using the generic dc on the main thread. r=rhunt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D19146
# ==============================================================================

diff -r 674fa918ee39 -r ab254a4faafe gfx/2d/DrawTargetD2D1.cpp
--- a/gfx/2d/DrawTargetD2D1.cpp	Sun Feb 10 11:50:49 2019 +0200
+++ b/gfx/2d/DrawTargetD2D1.cpp	Fri Feb 08 14:52:05 2019 +0100
@@ -1048,11 +1048,16 @@
     SurfaceFormat aFormat) const {
   RefPtr<ID2D1Bitmap1> bitmap;
 
-  HRESULT hr = Factory::GetD2DDeviceContext()->CreateBitmap(
-      D2DIntSize(aSize), aData, aStride,
-      D2D1::BitmapProperties1(D2D1_BITMAP_OPTIONS_NONE,
-                              D2DPixelFormat(aFormat)),
-      getter_AddRefs(bitmap));
+  RefPtr<ID2D1DeviceContext> dc = Factory::GetD2DDeviceContext();
+  if (!dc) {
+    return nullptr;
+  }
+
+  HRESULT hr =
+      dc->CreateBitmap(D2DIntSize(aSize), aData, aStride,
+                       D2D1::BitmapProperties1(D2D1_BITMAP_OPTIONS_NONE,
+                                               D2DPixelFormat(aFormat)),
+                       getter_AddRefs(bitmap));
 
   if (FAILED(hr) || !bitmap) {
     gfxCriticalError(
@@ -1062,8 +1067,8 @@
     return nullptr;
   }
 
-  return MakeAndAddRef<SourceSurfaceD2D1>(
-      bitmap.get(), Factory::GetD2DDeviceContext().get(), aFormat, aSize);
+  return MakeAndAddRef<SourceSurfaceD2D1>(bitmap.get(), dc.get(), aFormat,
+                                          aSize);
 }
 
 already_AddRefed<DrawTarget> DrawTargetD2D1::CreateSimilarDrawTarget(
@@ -1080,6 +1085,9 @@
 bool DrawTargetD2D1::CanCreateSimilarDrawTarget(const IntSize &aSize,
                                                 SurfaceFormat aFormat) const {
   RefPtr<ID2D1DeviceContext> dc = Factory::GetD2DDeviceContext();
+  if (!dc) {
+    return false;
+  }
   return (dc->GetMaximumBitmapSize() >= UINT32(aSize.width) &&
           dc->GetMaximumBitmapSize() >= UINT32(aSize.height));
 }
@@ -2162,6 +2170,11 @@
     return surface.forget();
   }
 
+  RefPtr<ID2D1DeviceContext> dc = Factory::GetD2DDeviceContext();
+  if (!dc) {
+    return nullptr;
+  }
+
   // Special case captures so we don't resolve them to a data surface.
   if (aSurface->GetType() == SurfaceType::CAPTURE) {
     SourceSurfaceCapture *capture =
@@ -2183,7 +2196,7 @@
       return nullptr;
     }
 
-    HRESULT hr = Factory::GetD2DDeviceContext()->CreateBitmap(
+    HRESULT hr = dc->CreateBitmap(
         D2DIntSize(data->GetSize()), map.GetData(), map.GetStride(),
         D2D1::BitmapProperties1(D2D1_BITMAP_OPTIONS_NONE,
                                 D2DPixelFormat(data->GetFormat())),
@@ -2201,8 +2214,7 @@
     return data.forget();
   }
 
-  return MakeAndAddRef<SourceSurfaceD2D1>(bitmap.get(),
-                                          Factory::GetD2DDeviceContext().get(),
+  return MakeAndAddRef<SourceSurfaceD2D1>(bitmap.get(), dc.get(),
                                           data->GetFormat(), data->GetSize());
 }
 
